### PR TITLE
Add compatibility-mode.md proposal.

### PR DESCRIPTION
**Pull request by SenorBlanco**
This proposal represents the Compatibility Mode issues which have achieved tentative consensus at the GPU for the Web Community Group. New issues will be added as they achieve consensus.

### CODE REVIEW

**Review by Kangz — APPROVED** [review]
LGTM overall. Will talk about this in today's meeting.

File: proposals/compatibility-mode.md
Comment (greggman) [reply proposals/compatibility-mode.md:57]: More apps will run with fewer changes like this

Comment (kainino0x) [reply proposals/compatibility-mode.md:57]: meeting: consensus to remove this special case, and match the defaulting rules that createView has.

File: proposals/compatibility-mode.md
Comment (SenorBlanco) [reply proposals/compatibility-mode.md:25]: I think the first part is covered in the "Goals" section above: "Since WebGPU Compatibility mode is a subset of WebGPU, all valid Compatibility mode applications are also valid WebGPU applications. Consequently, Compatibility mode applications will also run on user agents which do not support Compatibility mode."

I've added a sentence about reducing the restrictions to a minimum in the same section.

File: proposals/compatibility-mode.md
Comment (SenorBlanco) [reply proposals/compatibility-mode.md:57]: Modified to remove the 6-layer special case, as per CG consensus.

File: proposals/compatibility-mode.md
Comment (kainino0x) [inline proposals/compatibility-mode.md:50]: nit:
```suggestion
When specifying a texture, a `viewDimension` property determines the views which can be bound from that texture for sampling (see "Proposed IDL changes", above). Binding a view of a different dimension for sampling than specified at texture creation time will cause a validation error. If `viewDimension` is unspecified, use [the same algorithm as `createView()`](https://gpuweb.github.io/gpuweb/#abstract-opdef-resolving-gputextureviewdescriptor-defaults):
```

**Review by kainino0x — APPROVED** [review]
LGTM, if this is ready we should get PR approval from mwyrzykowski and jimblandy for formality.

**Review by mwyrzykowski — APPROVED** [review]

File: proposals/compatibility-mode.md
Comment (teoxoy) [reply proposals/compatibility-mode.md:25]: To clarify: do all implementations have to add the compat validation even if a non-compat adapter has been returned?

> Calls to APIs unsupported by Compatibility mode will result in validation errors.

I'm interpreting this as a "yes".

I think this is fine either way but it's worth clarifying since I don't remember this being discussed.

File: proposals/compatibility-mode.md
Comment (teoxoy) [inline proposals/compatibility-mode.md:50]:
```suggestion
When creating a texture, the `viewDimension` property determines the `dimension` of the view which can be created from that texture (see "Proposed IDL changes", above). Creating a view of a different dimension than specified at texture creation time will cause a validation error. If `viewDimension` is unspecified, use the following algorithm:
```

I'd clarify this is not specific to binding for sampling.

Comment (greggman) [reply proposals/compatibility-mode.md:50]: This is specific to using a texture as a TEXTURE_BINDING. In particular when you pass a texture view to `createBindGroup` is where this validation would apply.  For a RENDER_ATTACHMENT the viewDimension is not used. Instead, the existing non-compat validation suffice (https://www.w3.org/TR/webgpu/#abstract-opdef-renderable-texture-view)

Comment (greggman) [reply proposals/compatibility-mode.md:50]: To put it another way, you can create any view you can create in non-compat. It's only when you actually use the view does validation change from non-compat. The validation being specified here only happens on `createBindGroup`  Error if `view.dimension !== texture.viewDimension`

Comment (kainino0x) [reply proposals/compatibility-mode.md:50]: I had forgotten it happened in createBindGroup instead of createView. It makes sense though since texture views don't actually exist in GLES.

I wonder if `viewDimension` should be more clearly named? Like `textureBindingViewDimension`...

Comment (teoxoy) [reply proposals/compatibility-mode.md:50]: We should change the name if we validate this later but why not validate it earlier (at view creation)? I think it would make things clearer.

Comment (greggman) [reply proposals/compatibility-mode.md:50]: You can't validate a view creation because then you could not render to a cube map or a 2d array. To render to (to use as a RENDER_ATTACHMENT) requires that you use a '2d' dimension

```js
const pass = encoder.beginRenderPass({
    label: 'render to the positive Z face of a cube map',
    colorAttachments: [
      {
        view: myCubeMapTexture.createView({
          dimension: '2d',
          baseArrayLayer: 4,
          arrayLayerCount: 1,
          baseMipLevel: 0,
          mipLeveLCount: 1,
        }),
        ...
      },
    ],
});
```

vs, use as TEXTURE_BINDING cubemap

```js
    const bindGroup = device.createBindGroup({
      layout: pipeline.getBindGroupLayout(0),
      entries: [
        { binding: 0, resource: sampler },
        { binding: 1, resource: myCubeMapTexture.createView({dimension: 'cube'}) },
      ],
    });
```

Comment (teoxoy) [reply proposals/compatibility-mode.md:50]: I see, thanks for the clarification!

File: proposals/compatibility-mode.md
Comment (SenorBlanco) [reply proposals/compatibility-mode.md:25]: Depends what you mean by "non-compat adapter". If the developer requests compatibilityMode, then a supporting UA will return an Adapter that will validate all calls made on it against the Compatibility subset, even if the underlying physical device happens to be D3D12 or Vulkan, for example. Is that what you're asking?

(Nit: to me, such an Adapter is a "compat adapter", not a "non-compat adapter", since it appears to the developer to obey all the same constraints as one backed by OpenGL ES.)

File: proposals/compatibility-mode.md
Comment (SenorBlanco) [reply proposals/compatibility-mode.md]: I've left this as-is for now. Let me know if there are further edits desired here.

Comment (teoxoy) [reply proposals/compatibility-mode.md]: I still think it should be renamed; I'm ok with @kainino0x suggestion of `textureBindingViewDimension`.

File: proposals/compatibility-mode.md
Comment (teoxoy) [reply proposals/compatibility-mode.md:25]: So, when a user asks for `compatibilityMode = true` they can either get a "compat adapter" or a "non-compat adapter" depending on whether or not the UA supports the WebGPU Compat mode.

The problem I see is that there is no way for the user to know whether the UA supports the WebGPU Compat mode or not.
I see this being an issue for developers trying to test their compat apps on an UA that doesn't support compat and subsequently not getting any compat validation errors.

Comment (teoxoy) [reply proposals/compatibility-mode.md:25]: Was this written this way to make it easy to unship the compat mode in the future? If so, I guess this is fine but we should add a note somewhere in the spec that clarifies the fact that compat validation errors are not guaranteed on UAs that don't support the compat mode.

Comment (greggman) [reply proposals/compatibility-mode.md:25]: > I see this being an issue for developers trying to test their compat apps on an UA that doesn't support compat and subsequently not getting any compat validation errors.

It's not that difficult to provide a JavaScript extension or JS library that enforces most compat mode validation. I've written an incomplete one [here](https://github.com/greggman/webgpu-dev-extension/blob/main/extension/scripts/webgpu-compat-validation.js) and would be happy to complete it. Not sure if that helps this issue

Comment (SenorBlanco) [reply proposals/compatibility-mode.md:25]: Actually, I don't believe the initial motivation was related to unshipping. We were unsure whether all browser vendors would be willing to make Compatibility mode a required part of the WebGPU spec, or leave it optional. You're right that developers would be unable to test against compatibility mode if their browser did not support it. If all vendors are comfortable making it mandatory, we could reconsider the language above.

That said, I believe it is possible to check if a browser's support by checking the "isCompatibilityMode" property on the returned Adapter.

**Comment by SenorBlanco** [comment]
BTW, I linked my Github and W3C accounts, as per the "IPR contribution issue". However, the affiliation warning is still present. LMK if there's further action required here.

**Comment by kainino0x** [comment]
It took a couple of refreshes but somehow I made IPR revalidate correctly and it's happy now.

**Pull request merged**
